### ISSUE TITLE
Can't run inference: Could not load library libcublasLt.so.12. Error: libcublasLt.so.12: cannot open shared object file: No such file or directory

### ISSUE THREAD

**Issue by antortjim** [talmolab/sleap#2023]
# Bug description

The sleap GUI starts successfully, but SLEAP fails when I try to run inference on any frame(s) because one of the CUDA installation files appears to be missing (libcublasLt.so.12). I have found a potential solution (see below), so I leave it here in case anyone has the same issue.

## Expected behaviour

 I expected the inference to run smoothly and to get predictions in the desired frames.

## Your personal set up

- OS: Ubuntu 22.04

- Version(s):
-  SLEAP  v1.3.3
- Python 3.7.12
- CUDA 11 (I am not sure if SLEAP is using 11.3,7 or 8)
- Tensorflow 2.8.4

<details><summary>Logs</summary></details>

## Screenshots

![Image](https://github.com/user-attachments/assets/cdf43715-3121-4d5a-a6cb-0147cb94d16d)

## How to reproduce

1.
```
sleap-label labels.slp
```
2.  In the GUI: 
**Predict** > **Run Inference...** > Select the model you usually run analysis with (deployed) > **Run**

3. Wait a few seconds...

4.  In the terminal output I see:

INFO:sleap.nn.inference:Auto-selected GPU 0 with 48087 MiB of free memory.
2024-11-17 09:38:56.118572: W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:690] Error in PredictCost() for the op: op: "CropAndResize" attr { key: "T" value { type: DT_FLOAT } } attr { key: "extrapolation_value" value { f: 0 } } attr { key: "method" value { s: "bilinear" } } inputs { dtype: DT_FLOAT shape { dim { size: -49 } dim { size: -50 } dim { size: -51 } dim { size: 1 } } } inputs { dtype: DT_FLOAT shape { dim { size: -5 } dim { size: 4 } } } inputs { dtype: DT_INT32 shape { dim { size: -5 } } } inputs { dtype: DT_INT32 shape { dim { size: 2 } } } device { type: "GPU" vendor: "NVIDIA" model: "NVIDIA RTX A6000" frequency: 1800 num_cores: 84 environment { key: "architecture" value: "8.6" } environment { key: "cuda" value: "11020" } environment { key: "cudnn" value: "8100" } num_registers: 65536 l1_cache_size: 24576 l2_cache_size: 6291456 shared_memory_size_per_multiprocessor: 102400 memory_size: 48374874112 bandwidth: 768096000 } outputs { dtype: DT_FLOAT shape { dim { size: -5 } dim { size: -52 } dim { size: -53 } dim { size: 1 } } }
Could not load library libcublasLt.so.12. Error: libcublasLt.so.12: cannot open shared object file: No such file or directory

Process return code: -6

# Solution

This bug was probably caused in my system due to a system upgrade (since the same set up was working before). As the file name suggests, libcublasLt.so.12 belongs to a CUDA 12 installation, which is not supported by SLEAP. Going through the `environment.yml` file already indicates that tensorflow is pinned to `2.7.0` and I have a higher version which could be expecting the CUDA 12 file to be there. So there are two possible ways to fix this:
1) Follow the instructions from https://stackoverflow.com/questions/76646474/could-not-load-library-libcublaslt-so-12-error-libcublaslt-so-12-cannot-open#76739024
Basically, download the libcublaslt package, and place the lib files in the lib64 folder of your cuda installation. The files include the libcublasLt.so.12 file.
I have verified this solves the issue for me

2) Downgrade tensorflow to 2.7.0. I haven't verified this

**Comment by ValeriePineauNoel** [comment]
Hi @antortjim, 

Thank you for the insight. I am training SLEAP in Google Colab. Would you know how to solve this problem in Colab? I am trying to train a top-down pipeline. Training with `centroid.json` works fine; it's the training with `centered_instance.json` that gives me a similar error. Here is my notebook [20241121_TrainingSLEAPtopdown.ipynb.zip](https://github.com/user-attachments/files/17871292/20241121_TrainingSLEAPtopdown.ipynb.zip)

Thank you in advance your your help!

**Comment by ValeriePineauNoel** [comment]
Follow my last comment, I tried to download the libcublaslt package and place the lib files in the folder you suggested in colab. I did this by adding these lines at the end of the first cell, before importing sleap.
```
!wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/libcublaslt11_8.deb
!dpkg -i libcublaslt11_8.deb
!cp /usr/lib/x86_64-linux-gnu/libcublasLt.so.11_8 /usr/local/cuda/lib64/
!sudo ldconfig
!ldconfig
print("VERIFY THE INSTALLATION")
!ls /usr/local/cuda/lib64 | grep libcublasLt
```
Sadly, it did not work and gave me the same errors.

**Comment by antortjim** [comment]
Hi @ValeriePineauNoel , sorry I didn't reply earlier to your issue

I had a look at the notebook, and I see you attempt to install cuda 11.8. However, SLEAP has a requires CUDA 11.3.1 (see environment.yml file)

https://github.com/talmolab/sleap/blob/main/environment.yml
```
    # Packages required by tensorflow to find/use GPUs
    - conda-forge::cudatoolkit ==11.3.1
    # "==" results in package not found
    - conda-forge::cudnn=8.2.1
    - nvidia::cuda-nvcc=11.3
```

So if I were you I would try to stick to the CUDA version specified in the `environment.yml` file, unless you have a strong reason not to. It should be as easy as changing `apt install cuda-11-8` to `apt install cuda-11-3` (but can't guarantee). Good luck!

**Comment by eberrigan** [comment]
Hello @antortjim and @ValeriePineauNoel,

@antortjim our suggestion for your initial issue is to downgrade to `tensorflow` 2.7 because unfortunately `tensorflow` software updates result in backwards incompatibility. Our software tests have been done on specific versions of `tensorflow` in combination with specific versions of `cuda`. `Tensorflow` operations/commands vary a lot from version to version. 

@ValeriePineauNoel have you seen this issue #1644 ? We suggest running `!apt update && apt install cuda-11-8` before installing SLEAP. 
Thanks,

Elizabeth

**Comment by ValeriePineauNoel** [comment]
Hi @eberrigan,

I did follow issue [#1644 ](https://github.com/talmolab/sleap/issues/1644) at first. It worked well with training the bottom-up pipeline and training the first model of the top-down pipeline (with centroid.json). It's with training the second model of the top-down (centered_instance.json) that I have problems on Google Colab. Basically, the training starts, but stops around epoch 6, without doing the validation step, and I have this error displayed, which I thought looked like the error mentioned by @antortjim. 

```
[...]
2024-11-28 14:43:00.202853: W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:690] Error in PredictCost() for the op: op: "CropAndResize" attr { key: "T" value { type: DT_FLOAT } } attr { key: "extrapolation_value" value { f: 0 } } attr { key: "method" value { s: "bilinear" } } inputs { dtype: DT_FLOAT shape { dim { size: 1 } dim { size: 640 } dim { size: 640 } dim { size: 1 } } } inputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: 4 } } } inputs { dtype: DT_INT32 shape { dim { size: -2 } } } inputs { dtype: DT_INT32 shape { dim { size: 2 } } } device { type: "CPU" vendor: "GenuineIntel" model: "101" frequency: 2000 num_cores: 2 environment { key: "cpu_instruction_set" value: "AVX SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2" } environment { key: "eigen" value: "3.4.90" } l1_cache_size: 32768 l2_cache_size: 1048576 l3_cache_size: 40370176 memory_size: 268435456 } outputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: 656 } dim { size: 656 } dim { size: 1 } } }
Epoch 6/200
^C
```
